File: typescript/core/src/narc.ts
/**
 * Native secret-detection + prompt-injection scanning {@link ToolHook} — the
 * TypeScript port of the Rust reference engine's `narc.rs` (pearl th-5f7227).
 *
 * The SEP extension host passes tool-call arguments to the extension subprocess
 * **unscanned** and returns the subprocess's tool-result content to the model
 * **verbatim**. Nothing at the extension boundary looks for leaked credentials or
 * prompt-injection payloads. {@link NarcHook} closes that gap. It scans two things:
 *
 * - **Secrets** — 10 credential patterns (AWS keys, private keys, JWTs/bearer
 *   tokens, high-entropy provider keys, …).
 * - **Prompt injection** — 8 patterns (instruction override, role hijack,
 *   jailbreak, data/URL exfiltration, …).
 *
 * ## Division of labour with `PermissionHook`
 *
 * The permission gate already owns the *dangerous-command* / *write* /
 * *credential-path* circuit-breakers (`rm -rf /`, `curl | sh`, `~/.ssh/id_rsa`).
 * Narc does **not** re-implement those — it is scoped to the one thing permission
 * does not do: **content scanning of arguments and results** for secrets and
 * injection. Install Narc *after* the permission hook so the allow/ask/deny
 * decision happens first and Narc scans the calls that clear it.
 *
 * ## `preCall` (arguments) — blocks on exfiltration, alerts otherwise
 *
 * A {@link Severity.Block} injection match (the active data/URL exfiltration
 * signals) throws, blocking the call before the tool runs. Lower-severity
 * injection and any secret in the arguments are **alerted, not blocked** — a tool
 * argument legitimately carrying a secret (writing a `.env`, configuring a
 * client) is common enough that a hard block there would be a footgun.
 *
 * ## `postCall` (result) — detects, alerts, and **redacts** secrets
 *
 * `postCall` receives a **mutable** result, so a rewrite of `result.content` is
 * what downstream consumers — and the LLM/conversation — actually see. A secret
 * pattern in a tool result raises a {@link Severity.Block} alert *and* replaces
 * the matched credential with `[REDACTED:<pattern-name>]` before it reaches the
 * model. Injection patterns in the result remain detection + {@link Severity.Alert}
 * only (surveillance) — they can appear in legitimate content and are not rewritten.
 *
 * The detection set is pinned across all five engines by the shared corpus at
 * `spec/narc/corpus.json` (see `test/narc.test.ts`).
 */

import type { ToolCall, ToolHook, ToolResult } from './agent.js';

/**
 * Severity of a Narc finding, ordered least → most severe. A
 * {@link Severity.Block} finding in `preCall` blocks the tool call. The numeric
 * values ARE the ordering — compare with `>=`.
 */
export enum Severity {
    /** Informational — no action. */
    Info = 0,
    /** Suspicious but plausibly legitimate (e.g. a secret in an argument). */
    Warn = 1,
    /** Strong signal worth surfacing, but not auto-blocked. */
    Alert = 2,
    /** Actively harmful — blocks the call when raised in `preCall`. */
    Block = 3,
}

/** The shared wire label for a severity (`INFO`/`WARN`/`ALERT`/`BLOCK`). */
export function severityLabel(s: Severity): string {
    switch (s) {
        case Severity.Warn:
            return 'WARN';
        case Severity.Alert:
            return 'ALERT';
        case Severity.Block:
            return 'BLOCK';
        default:
            return 'INFO';
    }
}

/**
 * A single surveillance finding. Lean by design — the consumer supplies the
 * timestamp and correlation, so no uuid/timestamp fields are carried.
 */
export interface NarcAlert {
    /** How severe the finding is. */
    severity: Severity;
    /** Coarse bucket: `'injection'`, `'secret'`, `'secret_leak'`, `'injection_output'`. */
    category: string;
    /** The named pattern that matched. */
    patternName: string;
    /** Redacted view of the matched text (never the raw secret). */
    redacted: string;
    /** The tool whose args/result triggered the finding. */
    toolName: string;
}

/** A pattern match: which pattern, its severity, and a redacted view. */
export interface NarcFinding {
    /** The named pattern that matched. */
    patternName: string;
    /** The finding's severity. */
    severity: Severity;
    /** Redacted view of the matched text (safe to log). */
    redacted: string;
}

interface NamedPattern {
    name: string;
    severity: Severity;
    regex: RegExp;
}

/**
 * The 10 secret patterns. All are {@link Severity.Warn} in arguments (may be
 * legit) and escalate to {@link Severity.Block} when found in a result (a
 * leak) — the caller decides which threshold to apply.
 *
 * Every regex carries `g` because the scanners enumerate **all** matches. That
 * makes `RegExp.test` stateful on these objects, so this module never calls it —
 * `hasSecrets`/`hasInjection` go through the scanners.
 */
const SECRET_PATTERNS: NamedPattern[] = [
    { name: 'AWS Access Key', severity: Severity.Warn, regex: /AKIA[0-9A-Z]{16}/g },
    { name: 'AWS Secret Key', severity: Severity.Warn, regex: /aws[_\-]?secret[_\-]?access[_\-]?key\s*[=:]\s*[A-Za-z0-9/+=]{40}/gi },
    { name: 'Anthropic API Key', severity: Severity.Warn, regex: /sk-ant-[A-Za-z0-9\-_]{20,}/g },
    { name: 'OpenAI API Key', severity: Severity.Warn, regex: /sk-[A-Za-z0-9]{20,}/g },
    { name: 'GitHub Token', severity: Severity.Warn, regex: /gh[posr]_[A-Za-z0-9_]{36,}/g },
    { name: 'Private Key', severity: Severity.Warn, regex: /-----BEGIN\s+(RSA\s+)?PRIVATE\s+KEY-----/g },
    { name: 'Generic Secret', severity: Severity.Warn, regex: /(secret|password|token|api[_\-]?key)\s*[=:]\s*["']?[A-Za-z0-9/+=\-_]{8,}/gi },
    { name: 'Bearer Token', severity: Severity.Warn, regex: /Bearer\s+[A-Za-z0-9\-_.~+/]+=*/g },
    { name: 'Base64 Encoded Key', severity: Severity.Warn, regex: /(key|secret|password)\s*[=:]\s*[A-Za-z0-9+/]{32,}={0,2}/gi },
    { name: 'Stripe Key', severity: Severity.Warn, regex: /[sr]k_(live|test)_[A-Za-z0-9]{20,}/g },
];

/**
 * The 8 injection patterns. Only the active data/URL exfiltration signals are
 * {@link Severity.Block} (blocked in arguments); hijack/jailbreak text is
 * {@link Severity.Alert} (surveilled, not blocked — it can appear in legitimate
 * content the model is authoring, e.g. a security test or documentation about
 * injection).
 *
 * `data_exfiltration` is the Rust pattern's free-spacing `(?x)` form flattened to
 * one line: JavaScript has no `x` flag, and a rewrite that dropped an alternative
 * would silently weaken the detector.
 */
const INJECTION_PATTERNS: NamedPattern[] = [
    { name: 'ignore_instructions', severity: Severity.Alert, regex: /ignore\s+(all\s+)?(previous|prior|above)\s+(instructions|prompts|rules)/gi },
    { name: 'role_hijack', severity: Severity.Alert, regex: /(you\s+are\s+now|act\s+as|pretend\s+(to\s+be|you\s+are)|from\s+now\s+on\s+you\s+are)/gi },
    { name: 'system_prompt', severity: Severity.Alert, regex: /(system\s*:\s*|<\|system\|>|\[SYSTEM\])/gi },
    { name: 'jailbreak', severity: Severity.Alert, regex: /(DAN\s+mode|developer\s+mode|do\s+anything\s+now|jailbreak)/gi },
    { name: 'base64_smuggling', severity: Severity.Alert, regex: /(decode|eval|execute)\s+(this\s+)?(base64|encoded)/gi },
    {
        name: 'data_exfiltration',
        severity: Severity.Block,
        regex: /(send|post|upload|exfiltrate|transmit|leak|push)\s+(all\s+|the\s+|our\s+|my\s+|this\s+)*(data|files?|secrets?|credentials?|keys?|tokens?|contents?|env\s+(vars?|file)|package\.json|\.env|pyproject\.toml|cargo\.toml|requirements\.txt|gemfile|go\.mod|composer\.json|\.ssh\/[a-z_]+|id_rsa|\.aws\/[a-z]+|\.gnupg\/)\s+(to|via|at|over)/gi,
    },
    {
        name: 'url_exfiltration',
        severity: Severity.Block,
        regex: /(send|post|upload|push|transmit|leak|exfiltrate)\b[^.\n]{1,200}\s+(to|via|at|over)\s+(https?:\/\/[\w.\-/]+)/gi,
    },
    { name: 'smell_url', severity: Severity.Alert, regex: /https?:\/\/[\w.\-]*\b(leak|exfil|attacker|evil|tracker|c2(?:server)?|webhook\.site)\b[\w.\-/]*/gi },
];

function scan(patterns: NamedPattern[], text: string): NarcFinding[] {
    const out: NarcFinding[] = [];
    for (const p of patterns) {
        for (const m of text.matchAll(p.regex)) {
            out.push({ patternName: p.name, severity: p.severity, redacted: redactMatch(m[0]) });
        }
    }
    return out;
}

/** Scan `text` for hardcoded secrets. Every match is redacted. */
export function scanSecrets(text: string): NarcFinding[] {
    return scan(SECRET_PATTERNS, text);
}

/** Scan `text` for prompt-injection patterns. Matched text is redacted. */
export function scanInjection(text: string): NarcFinding[] {
    return scan(INJECTION_PATTERNS, text);
}

/** True if `text` contains any secret pattern. */
export function hasSecrets(text: string): boolean {
    return scanSecrets(text).length > 0;
}

/** True if `text` contains any injection pattern. */
export function hasInjection(text: string): boolean {
    return scanInjection(text).length > 0;
}

/**
 * Redact a matched string, showing only the first 4 and last 2 characters. Short
 * matches (≤ 8 code points) are fully starred. Code points, not UTF-16 units —
 * parity with the Rust reference's char-based redaction.
 */
export function redactMatch(s: string): string {
    const chars = [...s];
    if (chars.length <= 8) return '*'.repeat(chars.length);
    return chars.slice(0, 4).join('') + '*'.repeat(chars.length - 6) + '**' + chars.slice(chars.length - 2).join('');
}

/**
 * A {@link ToolHook} that scans tool-call arguments and results for secrets and
 * prompt injection. Install it via `AgentOptions.toolHooks` alongside the
 * permission hook, *after* it.
 *
 * - **`preCall`** throws on a {@link Severity.Block} injection pattern in the
 *   arguments (active exfiltration); every other finding (lower-severity
 *   injection, any secret) is recorded as a {@link NarcAlert}, not blocked.
 * - **`postCall`** detects secrets/injection in the result, records them, and
 *   **redacts** leaked secrets out of the content in place so the model never
 *   sees the raw credential.
 */
export class NarcHook implements ToolHook {
    private readonly log: NarcAlert[] = [];

    /** Snapshot every recorded alert. */
    alerts(): NarcAlert[] {
        return [...this.log];
    }

    /** Recorded alerts at or above `minSeverity`. */
    alertsAbove(minSeverity: Severity): NarcAlert[] {
        return this.log.filter((a) => a.severity >= minSeverity);
    }

    private record(alert: NarcAlert): void {
        this.log.push(alert);
    }

    async preCall(call: ToolCall): Promise<void> {
        const argsText = JSON.stringify(call.arguments);

        // Scan all first so every finding is recorded even when one of them blocks.
        let block: NarcFinding | undefined;
        for (const f of scanInjection(argsText)) {
            if (f.severity >= Severity.Block && block === undefined) block = f;
            this.record({ severity: f.severity, category: 'injection', patternName: f.patternName, redacted: f.redacted, toolName: call.name });
        }

        // Secrets in arguments: alert only (may be legitimate).
        for (const f of scanSecrets(argsText)) {
            this.record({ severity: f.severity, category: 'secret', patternName: f.patternName, redacted: f.redacted, toolName: call.name });
        }

        if (block !== undefined) {
            throw new Error(`prompt-injection pattern \`${block.patternName}\` in tool arguments — blocked`);
        }
    }

    async postCall(call: ToolCall, result: ToolResult): Promise<void> {
        // A secret in a tool result is a leak. Record a Block alert AND redact it
        // out of `result.content` — the mutable seam means this rewrite is what
        // the model/conversation and every downstream consumer actually see.
        const secrets = scanSecrets(result.content);
        for (const f of secrets) {
            this.record({ severity: Severity.Block, category: 'secret_leak', patternName: f.patternName, redacted: f.redacted, toolName: call.name });
        }
        if (secrets.length > 0) {
            let content = result.content;
            for (const p of SECRET_PATTERNS) {
                // Replace via a function so `$&`-style sequences in a pattern name
                // could never be interpolated into the output.
                content = content.replace(p.regex, () => `[REDACTED:${p.name}]`);
            }
            result.content = content;
        }
        // Injection in the result is detection + alert only (surveillance) — scan
        // the post-redaction content the model will actually see.
        for (const f of scanInjection(result.content)) {
            this.record({
                severity: f.severity > Severity.Alert ? f.severity : Severity.Alert,
                category: 'injection_output',
                patternName: f.patternName,
                redacted: f.redacted,
                toolName: call.name,
            });
        }
    }
}

File: go/core/narc.go
package core

// Native secret-detection + prompt-injection scanning ToolHook — the Go port of
// the Rust reference engine's `narc.rs` (pearl th-5f7227).
//
// The SEP ExtensionHost passes tool-call arguments to the extension subprocess
// UNSCANNED and returns the subprocess's tool-result content to the model
// VERBATIM. Nothing at the extension boundary looks for leaked credentials or
// prompt-injection payloads. [NarcHook] closes that gap. It scans two things:
//
//   - Secrets — 10 credential patterns (AWS keys, private keys, JWTs/bearer
//     tokens, high-entropy provider keys, …).
//   - Prompt injection — 8 patterns (instruction override, role hijack,
//     jailbreak, data/URL exfiltration, …).
//
// Division of labour with [PermissionGate]: the permission gate owns the
// dangerous-command / write / credential-path circuit-breakers (rm -rf /,
// curl | sh, ~/.ssh/id_rsa). Narc does NOT re-implement those — it is scoped to
// the one thing permission does not do: CONTENT SCANNING of arguments and
// results for secrets and injection. Install Narc AFTER the permission gate so
// the allow/ask/deny decision happens first and Narc scans what clears it.
//
// PreCall (arguments) — blocks on exfiltration, alerts otherwise. A
// [SeverityBlock] injection match (the active data/URL exfiltration signals)
// returns an error, blocking the call before it reaches the tool. Lower-severity
// injection and ANY secret in the arguments are alerted, not blocked — a tool
// argument legitimately carrying a secret (writing a .env, configuring a client)
// is common enough that a hard block there would be a footgun.
//
// PostCall (result) — detects, alerts, and REDACTS secrets. PostCall takes a
// *ToolResult, so a hook's rewrite of Content is what downstream consumers — and
// the LLM/conversation — actually see. A secret pattern in a tool result raises a
// [SeverityBlock] alert AND replaces the matched credential with
// `[REDACTED:<pattern-name>]` in the content before it reaches the model.
// Injection patterns in the result remain detection + [SeverityAlert] only
// (surveillance) — they can appear in legitimate content and are not rewritten.
//
// The detection set is pinned across all five engines by the shared corpus at
// spec/narc/corpus.json (see narc_test.go).

import (
	"context"
	"fmt"
	"log"
	"regexp"
	"strings"
	"sync"
)

// Severity is how severe a Narc finding is, ordered least → most severe. A
// [SeverityBlock] finding in PreCall blocks the tool call.
type Severity int

const (
	// SeverityInfo is informational — no action.
	SeverityInfo Severity = iota
	// SeverityWarn is suspicious but plausibly legitimate (e.g. a secret in an argument).
	SeverityWarn
	// SeverityAlert is a strong signal worth surfacing, but not auto-blocked.
	SeverityAlert
	// SeverityBlock is actively harmful — blocks the call when raised in PreCall.
	SeverityBlock
)

// String renders the severity as the shared wire label (INFO/WARN/ALERT/BLOCK).
func (s Severity) String() string {
	switch s {
	case SeverityInfo:
		return "INFO"
	case SeverityWarn:
		return "WARN"
	case SeverityAlert:
		return "ALERT"
	case SeverityBlock:
		return "BLOCK"
	default:
		return "INFO"
	}
}

// NarcAlert is a single surveillance finding. Lean by design — the log line
// supplies the timestamp, so no uuid/timestamp fields are carried.
type NarcAlert struct {
	// Severity is how severe the finding is.
	Severity Severity
	// Category is the coarse bucket: "injection", "secret", "secret_leak", "injection_output".
	Category string
	// PatternName is the named pattern that matched.
	PatternName string
	// Redacted is a redacted view of the matched text (never the raw secret).
	Redacted string
	// ToolName is the tool whose args/result triggered the finding.
	ToolName string
}

// NarcFinding is a pattern match: which pattern, its severity, and a redacted view.
type NarcFinding struct {
	// PatternName is the named pattern that matched.
	PatternName string
	// Severity is the finding's severity.
	Severity Severity
	// Redacted is a redacted view of the matched text (safe to log).
	Redacted string
}

// narcPattern is one named detector.
type narcPattern struct {
	name     string
	severity Severity
	re       *regexp.Regexp
}

// secretPatterns is the 10 credential patterns. All are [SeverityWarn] in
// arguments (may be legit) and escalate to [SeverityBlock] when found in a
// result (a leak) — the caller decides which threshold to apply.
var secretPatterns = []narcPattern{
	{"AWS Access Key", SeverityWarn, regexp.MustCompile(`AKIA[0-9A-Z]{16}`)},
	{"AWS Secret Key", SeverityWarn, regexp.MustCompile(`(?i)aws[_\-]?secret[_\-]?access[_\-]?key\s*[=:]\s*[A-Za-z0-9/+=]{40}`)},
	{"Anthropic API Key", SeverityWarn, regexp.MustCompile(`sk-ant-[A-Za-z0-9\-_]{20,}`)},
	{"OpenAI API Key", SeverityWarn, regexp.MustCompile(`sk-[A-Za-z0-9]{20,}`)},
	{"GitHub Token", SeverityWarn, regexp.MustCompile(`gh[posr]_[A-Za-z0-9_]{36,}`)},
	{"Private Key", SeverityWarn, regexp.MustCompile(`-----BEGIN\s+(RSA\s+)?PRIVATE\s+KEY-----`)},
	{"Generic Secret", SeverityWarn, regexp.MustCompile(`(?i)(secret|password|token|api[_\-]?key)\s*[=:]\s*["']?[A-Za-z0-9/+=\-_]{8,}`)},
	{"Bearer Token", SeverityWarn, regexp.MustCompile(`Bearer\s+[A-Za-z0-9\-_.~+/]+=*`)},
	{"Base64 Encoded Key", SeverityWarn, regexp.MustCompile(`(?i)(key|secret|password)\s*[=:]\s*[A-Za-z0-9+/]{32,}={0,2}`)},
	{"Stripe Key", SeverityWarn, regexp.MustCompile(`[sr]k_(live|test)_[A-Za-z0-9]{20,}`)},
}

// injectionPatterns is the 8 prompt-injection patterns. Only the active
// data/URL exfiltration signals are [SeverityBlock] (blocked in arguments);
// hijack/jailbreak text is [SeverityAlert] (surveilled, not blocked — it can
// appear in legitimate content the model is authoring, e.g. a security test or
// documentation about injection).
//
// data_exfiltration is the Rust pattern's free-spacing `(?x)` form flattened to
// one line: Go's RE2 has no `x` flag, and a rewrite that dropped an alternative
// would silently weaken the detector.
var injectionPatterns = []narcPattern{
	{"ignore_instructions", SeverityAlert, regexp.MustCompile(`(?i)ignore\s+(all\s+)?(previous|prior|above)\s+(instructions|prompts|rules)`)},
	{"role_hijack", SeverityAlert, regexp.MustCompile(`(?i)(you\s+are\s+now|act\s+as|pretend\s+(to\s+be|you\s+are)|from\s+now\s+on\s+you\s+are)`)},
	{"system_prompt", SeverityAlert, regexp.MustCompile(`(?i)(system\s*:\s*|<\|system\|>|\[SYSTEM\])`)},
	{"jailbreak", SeverityAlert, regexp.MustCompile(`(?i)(DAN\s+mode|developer\s+mode|do\s+anything\s+now|jailbreak)`)},
	{"base64_smuggling", SeverityAlert, regexp.MustCompile(`(?i)(decode|eval|execute)\s+(this\s+)?(base64|encoded)`)},
	{"data_exfiltration", SeverityBlock, regexp.MustCompile(`(?i)(send|post|upload|exfiltrate|transmit|leak|push)\s+(all\s+|the\s+|our\s+|my\s+|this\s+)*(data|files?|secrets?|credentials?|keys?|tokens?|contents?|env\s+(vars?|file)|package\.json|\.env|pyproject\.toml|cargo\.toml|requirements\.txt|gemfile|go\.mod|composer\.json|\.ssh/[a-z_]+|id_rsa|\.aws/[a-z]+|\.gnupg/)\s+(to|via|at|over)`)},
	{"url_exfiltration", SeverityBlock, regexp.MustCompile(`(?i)(send|post|upload|push|transmit|leak|exfiltrate)\b[^.\n]{1,200}\s+(to|via|at|over)\s+(https?://[\w.\-/]+)`)},
	{"smell_url", SeverityAlert, regexp.MustCompile(`(?i)https?://[\w.\-]*\b(leak|exfil|attacker|evil|tracker|c2(?:server)?|webhook\.site)\b[\w.\-/]*`)},
}

// scan runs every pattern over text and returns one finding per match, in
// pattern order.
func scan(patterns []narcPattern, text string) []NarcFinding {
	out := []NarcFinding{}
	for _, p := range patterns {
		for _, m := range p.re.FindAllString(text, -1) {
			out = append(out, NarcFinding{PatternName: p.name, Severity: p.severity, Redacted: RedactMatch(m)})
		}
	}
	return out
}

// ScanSecrets scans text for hardcoded secrets. Every match is redacted.
func ScanSecrets(text string) []NarcFinding { return scan(secretPatterns, text) }

// ScanInjection scans text for prompt-injection patterns. Matched text is redacted.
func ScanInjection(text string) []NarcFinding { return scan(injectionPatterns, text) }

// HasSecrets reports whether text contains any secret pattern.
func HasSecrets(text string) bool {
	for _, p := range secretPatterns {
		if p.re.MatchString(text) {
			return true
		}
	}
	return false
}

// HasInjection reports whether text contains any injection pattern.
func HasInjection(text string) bool {
	for _, p := range injectionPatterns {
		if p.re.MatchString(text) {
			return true
		}
	}
	return false
}

// RedactMatch redacts a matched string, showing only the first 4 and last 2
// characters. Short matches (≤ 8 runes) are fully starred. Runes, not bytes —
// parity with the Rust reference's char-based redaction.
func RedactMatch(s string) string {
	r := []rune(s)
	if len(r) <= 8 {
		return strings.Repeat("*", len(r))
	}
	return string(r[:4]) + strings.Repeat("*", len(r)-6) + "**" + string(r[len(r)-2:])
}

// NarcHook is a [ToolHook] that scans tool-call arguments and results for
// secrets and prompt injection. Install it on AgentOptions.Hooks alongside the
// permission gate, AFTER it, so the permission gate decides allow/ask/deny first
// and Narc scans the calls that clear it.
//
//   - PreCall blocks on a [SeverityBlock] injection pattern in the arguments
//     (active exfiltration); every other finding (lower-severity injection, any
//     secret) is recorded as a [NarcAlert] and logged, not blocked.
//   - PostCall detects secrets/injection in the result, records + logs them, and
//     REDACTS leaked secrets out of the content in place so the model never sees
//     the raw credential.
//
// Safe for concurrent use: with ParallelToolCalls the engine may run
// PreCall/PostCall from several goroutines at once.
type NarcHook struct {
	mu     sync.Mutex
	alerts []NarcAlert
}

// NewNarcHook builds a fresh hook with an empty alert log.
func NewNarcHook() *NarcHook { return &NarcHook{} }

// Alerts snapshots every recorded alert.
func (h *NarcHook) Alerts() []NarcAlert {
	h.mu.Lock()
	defer h.mu.Unlock()
	return append([]NarcAlert(nil), h.alerts...)
}

// AlertsAbove returns the recorded alerts at or above minSeverity.
func (h *NarcHook) AlertsAbove(minSeverity Severity) []NarcAlert {
	h.mu.Lock()
	defer h.mu.Unlock()
	out := []NarcAlert{}
	for _, a := range h.alerts {
		if a.Severity >= minSeverity {
			out = append(out, a)
		}
	}
	return out
}

func (h *NarcHook) record(a NarcAlert) {
	log.Printf("narc: %s finding tool=%s category=%s pattern=%s redacted=%s",
		a.Severity, a.ToolName, a.Category, a.PatternName, a.Redacted)
	h.mu.Lock()
	defer h.mu.Unlock()
	h.alerts = append(h.alerts, a)
}

// PreCall scans the tool arguments. A Block-severity injection pattern (active
// exfiltration) returns an error and the tool never runs; everything else is
// recorded and allowed.
func (h *NarcHook) PreCall(_ context.Context, call ToolCall) error {
	args := call.Arguments

	// Scan all first so every finding is recorded even when one of them blocks.
	var block *NarcFinding
	for _, f := range ScanInjection(args) {
		if f.Severity >= SeverityBlock && block == nil {
			hit := f
			block = &hit
		}
		h.record(NarcAlert{Severity: f.Severity, Category: "injection", PatternName: f.PatternName, Redacted: f.Redacted, ToolName: call.Name})
	}

	// Secrets in arguments: alert only (may be legitimate).
	for _, f := range ScanSecrets(args) {
		h.record(NarcAlert{Severity: f.Severity, Category: "secret", PatternName: f.PatternName, Redacted: f.Redacted, ToolName: call.Name})
	}

	if block != nil {
		return fmt.Errorf("prompt-injection pattern %q in tool arguments — blocked", block.PatternName)
	}
	return nil
}

// PostCall scans the tool result. A secret in a result is a leak: it raises a
// Block alert AND is redacted out of result.Content in place. Injection in the
// result is surveillance only — recorded, never rewritten.
func (h *NarcHook) PostCall(_ context.Context, call ToolCall, result *ToolResult) error {
	secrets := ScanSecrets(result.Content)
	for _, f := range secrets {
		h.record(NarcAlert{Severity: SeverityBlock, Category: "secret_leak", PatternName: f.PatternName, Redacted: f.Redacted, ToolName: call.Name})
	}
	if len(secrets) > 0 {
		content := result.Content
		for _, p := range secretPatterns {
			content = p.re.ReplaceAllLiteralString(content, "[REDACTED:"+p.name+"]")
		}
		result.Content = content
	}
	// Scan the POST-redaction content — what the model will actually see.
	for _, f := range ScanInjection(result.Content) {
		sev := f.Severity
		if sev < SeverityAlert {
			sev = SeverityAlert
		}
		h.record(NarcAlert{Severity: sev, Category: "injection_output", PatternName: f.PatternName, Redacted: f.Redacted, ToolName: call.Name})
	}
	return nil
}
